### PR TITLE
Add packaging guidelines

### LINKED ISSUE #352: Packaging guidelines
I think it would make sense to document a few things for packaging vdirsyncer.

## How to test vdirsyncer

Many packagers try to run the full testsuite on a vdirsyncer installation, in order to find bugs with their setup. But: 99% of all packaging-related bugs are caught by seeing if `vdirsyncer --help` works correctly.

The testsuite is not suitable for packaging tests because:

* It uses [hypothesis](http://hypothesis.readthedocs.org/) to randomly generate example input data. Automatic test generation is a really powerful tool for finding encoding-related bugs, but it also means that tests may randomly fail.

* It launches CardDAV/CalDAV servers for integration testing. In other words, it's not just for finding bugs in vdirsyncer, but also elsewhere. Why would vdirsyncer-packagers care if a server introduced a bug that makes the tests fail?

* It requires very recent dependencies. Older pytest, hypothesis or Radicale versions may fail with bizzarre errors. This is especially a problem if packagers try to use their system-packages for installing test dependencies instead of just using a virtualenv and pip (I don't see a reason why that would be a problem TBH) See also #334 

This seems to be the sanest approach to testing vdirsyncer as a packager: https://github.com/Homebrew/homebrew/pull/41208 i.e. just create an example config and run a simple sync. Even a simple `vdirsyncer --help` catches 80% of errors IMO.

## Updating vdirsyncer

[The Hypothesis Packaging Guidelines](https://hypothesis.readthedocs.org/en/release/packaging.html) say that package maintainers who are unable to keep up with its release cycle are doing the community a disservice. While I think there are nicer ways to phrase that (especially towards people who are likewise doing a voluntary service), I am worried that vdirsyncer will find itself in a similarly problematic situation once [Debian starts to package it](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=745027). Packages usually have been very quickly updated otherwise.

Apart from that we should probably provide an RSS feed for new releases. GitHub has a tags RSS feed that is mentioned [in the docs](https://vdirsyncer.readthedocs.org/en/latest/changelog.html), I guess we could just use that.

## Enforcing dependencies' minimum required versions

Vdirsyncer has a lower bound on the versions of some dependencies. As far as I understand those lower bounds have been completely ignored so far by downstream. Perhaps there's just no way to fix this situation. Again I'm more worried about Debian than any other distro here, but maybe it'll be fine.

--

Seeking feedback from

* @hobarrera @eolianoe (ArchLinux)
* @0-wiz-0 (pkgsrc)
* @mbaldessari @mathstuf (Fedora)
* @damiencassou @matthiasbeyer (nix)
* @benjaminfrank (Homebrew)
* @lfam (GNU + Guix)

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31504145-packaging-guidelines?utm_campaign=plugin&utm_content=tracker%2F574345&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F574345&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by untitaker**
Fix #352

### CODE REVIEW

**Comment by hobarrera** [comment]
Thanks for this! I wish more upstream developers took the time to provide this sort of documentation to downstream packagers!

**Comment by DamienCassou** [comment]
Thank you very much. Where is `dev-requirements.txt`?

File: docs/packaging.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+Dependency versions
+===================
+
+It is strongly discouraged to package vdirsyncer as a Python 2 application.
+Future releases will only work on Python 3.3 and newer versions.
+
+As with most Python packages, ``setup.py`` denotes the runtime dependencies of
+vdirsyncer. It also contains lower-bound versions of each dependency. Older
+versions may appear to work (since e.g. all tests pass), but they really don't.
```
Comment (DamienCassou) [inline docs/packaging.rst]: is it possible for you to write a unit test that fails if a lower bound is not satisfied? You would get less bug reports I guess.

**Comment by eolianoe** [comment]
Thanks for those indications about all the dependencies.

It sees that werkzeug is missing from [test-requirements.txt](https://github.com/untitaker/vdirsyncer/blob/922424ff4431bc26b29dd3bd5346d84f0a924525/test-requirements.txt) as the tests are failing if it is not installed:
```
__________________________________ ERROR at setup of test_request_ssl __________________________________
file vdirsyncer/tests/utils/test_main.py, line 44
  def test_request_ssl(httpsserver):
        fixture 'httpsserver' not found
        available fixtures: pytestconfig, tmpdir_factory, setup_logging, capfd, record_xml_property, subtest, monkeypatch, cache, suppress_py2_warning, no_debug_output, benchmark, tmpdir, recwarn, capsys
        use 'py.test --fixtures [testpath]' for help on them.

vdirsyncer/tests/utils/test_main.py:44
 ERROR at setup of test_request_ssl_fingerprints[94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA] 
file vdirsyncer/tests/utils/test_main.py, line 63
  @pytest.mark.skipif(_fingerprints_broken(),
                      reason='https://github.com/shazow/urllib3/issues/529')
  @pytest.mark.parametrize('fingerprint', [
      '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA',
      '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D'
  ])
  def test_request_ssl_fingerprints(httpsserver, fingerprint):
        fixture 'httpsserver' not found
        available fixtures: pytestconfig, tmpdir_factory, setup_logging, capfd, record_xml_property, subtest, monkeypatch, cache, suppress_py2_warning, no_debug_output, benchmark, tmpdir, recwarn, capsys
        use 'py.test --fixtures [testpath]' for help on them.

vdirsyncer/tests/utils/test_main.py:63
___ ERROR at setup of test_request_ssl_fingerprints[19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D] ___
file vdirsyncer/tests/utils/test_main.py, line 63
  @pytest.mark.skipif(_fingerprints_broken(),
                      reason='https://github.com/shazow/urllib3/issues/529')
  @pytest.mark.parametrize('fingerprint', [
      '94:FD:7A:CB:50:75:A4:69:82:0A:F8:23:DF:07:FC:69:3E:CD:90:CA',
      '19:90:F7:23:94:F2:EF:AB:2B:64:2D:57:3D:25:95:2D'
  ])
  def test_request_ssl_fingerprints(httpsserver, fingerprint):
        fixture 'httpsserver' not found
        available fixtures: pytestconfig, tmpdir_factory, setup_logging, capfd, record_xml_property, subtest, monkeypatch, cache, suppress_py2_warning, no_debug_output, benchmark, tmpdir, recwarn, capsys
        use 'py.test --fixtures [testpath]' for help on them.

vdirsyncer/tests/utils/test_main.py:63
```

**Comment by untitaker** [comment]
Afaik pytest-localserver already depends on it?

**Comment by untitaker** [comment]
@DamienCassou Sorry, I meant `test-requirements.txt` and failed to rename. This is fixed now.

File: docs/packaging.rst (hunk 1 above)
Comment (untitaker) [inline docs/packaging.rst]: Done.

**Comment by untitaker** [comment]
I would merge this PR as-is, and do the separation of Hypothesis tests another time, as it's not that easy. @DamienCassou you'll have to skip the whole testsuite for now, but IMO the testsuite is deterministic enough as-is in practice. Hypothesis tests have been in vdirsyncer for quite a while now and nobody has noticed.

**Comment by DamienCassou** [comment]
Please merge. I will see if I can activate the tests nonetheless. If I see failures, I will come back and haunt you :-).

**Comment by hobarrera** [comment]
LGTM 👍

**Comment by untitaker** [comment]
@eolianoe BTW the problems with Werkzeug missing were a problem with the `python{,2}-pytest-localserver` packages in AUR.

**Comment by eolianoe** [comment]
@untitaker: yes, I saw the changes in the AUR package. Thanks @hobarrera for that

**Pull request merged**